### PR TITLE
fix ordering of consumer/publisher delivery tags

### PR DESCRIPTION
**Pull request by embray**
This fixes a tricky bug which can occur especially if there are two
open channels on the same connection, though it might be possible to
reproduce even on a single channel (needs a simpler test if we can
figure out how).

The main issue is that the delivery tags used for publisher confirmation
and delivery confirmation are *not* the same, and should use different
sequences.

For example when a client publishes a message with delivery confirmation
expected, it waits for a (n)ack with the matching delivery tag.  That
means the mock server needs to send a (n)ack with the correct delivery
tag.

However, if the client is acting as a consumer, and server is expecting
delivery confirmation, then the server must exect a (n)ack from the
client with a matching delivery_tag.

In other words, the server needs to keep track of two separate
sequences of delivery tags: one where the server is acting as a
publisher, and one where it's handling a consumer.  The delivery tags
are also channel-specific, so this adds keying by channel for the
delivery tags.

(NOTE: I wrote this fix many months ago but never wrote up a proper
description of it until now, so I'm not 100% sure my description of the
problem is accurate.)

Depends on https://github.com/nikitanovosibirsk/amqp-mock/pull/8 and https://github.com/nikitanovosibirsk/amqp-mock/pull/9

### CODE REVIEW

**Comment by embray** [comment]
Rebased (still includes #9)

**Comment by skrech** [comment]
I'm not sure how AMQP protocol defines this, but what is the reason to count publisher confirms? As I see things, counting should happen only on the sending side, eg.:
1. Broker counts messages sent to consumers (and consumers just respond with the delivery_id that they receive in the `delivery` headers)
2. Clients count messages sent to broker (and broker just returns confirmations based on what theu receive in the `publish` headers.
